domain/network/state/linklayer_merge.go

```
@@ -229,22 +211,115 @@ func (st *State) applyMergeLinkLayerChanges(
 		return errors.Errorf("relinquishing addresses: %w", err)
 	}
 
+	// Process subnet updates
+	err = st.updateSubnets(ctx, tx, addressChanges.subnetToUpdate)
+	if err != nil {
+		return errors.Errorf("updating subnets: %w", err)
+	}
+
+	// Remove subnets that are no longer needed
+	if err := st.cleanupUniqueAddressOrphanSubnets(ctx, tx); err != nil {
+		return errors.Errorf("cleaning up orphan subnets: %w", err)
+	}
```

question: Comments below indicate that a failure to cleanup a subnet is okay for now if there is a provider_id. But here you're failing the entire transaction which includes `applyMergeLinkLayerChanges` on this point. Is that appropriate?

---

I think so, because even if the `removeSubnet` method will fail if the subnet is still linked to a provider_id, it is called after a select which get subnet without provider_id.

So the whole stack is safe, and if an error occurs, it won't be this one.

If for some reason it happens (for instance the clause is removded from the query), it is a bug introduced through dev, and should fail fast. 

I added two test case to make sure such change would be properly catched by TU

---

**todo:** provide better comments